File: javascript-sdk/packages/core/src/types/client-response.ts
import { I18n } from "../i18n";

export enum ClientCode {
  success = 'success',
  payment_success = 'payment_success',
  payment_failed = 'payment_failed',
  payment_partial = 'payment_partial',
  payment_timeout = 'payment_timeout',
  connect_error = 'connect_error',
  socket_connection_error = 'socket_connection_error',
  data_load_error = 'data_load_error',
  qr_fetch_error = 'qr_fetch_error',
  paymentProcessNotAvailable = 'payment_process_not_available',
  get_assets_error = 'get_assets_error',
  unknown_error = 'unknown_error'
}

export abstract class ClientResponseBase {
  readonly code: ClientCode;
  readonly message: string;

  constructor({ code, message }: { code: ClientCode, message: string }) {
    this.code = code;
    this.message = message;
  }
}

export class ClientResponse<T = any> extends ClientResponseBase {
  readonly result?: T;

  constructor({ code, message, result }: { code: ClientCode, message: string, result?: T }) {
    super({ code, message });
    this.result = result;
  }

  static fromJson<T = any>(json: any, { code, message, result }: {
    code?: ClientCode,
    message?: string,
    result?: (json: any) => T
  } = {}): ClientResponse<T> {
    const successCode = json.status || code || ClientCode.success;
    const successMessage = json.message || message || I18n.t.successes.success;
    const successResult = result?.(json.result) || json.result || json;

    return new ClientResponse<T>({ code: successCode, message: successMessage, result: successResult });
  }
}

export class ClientError extends ClientResponseBase {
  readonly error?: any;

  constructor({ code, message, error }: { code: ClientCode, message: string, error?: any }) {
    super({ code, message });
    this.error = error;
  }

  static fromError(error: any, { code, message }: { code?: ClientCode, message?: string } = {}): ClientError {
    if (error instanceof ClientError) return error;

    const errCode = code || error.statusCode || ClientCode.unknown_error;
    const errMessage = message || error.message || I18n.t.errors.unknownError;
    const err = error.error || error;

    return new ClientError({ code: errCode, message: errMessage, error: err });
  }
}

File: dart_sdk/lib/models/client_response.dart
import 'package:apolopay_sdk/apolopay_sdk.dart';

enum ClientCode {
  success('success'),
  paymentSuccess('payment_success'),
  paymentFailed('payment_failed'),
  paymentPartial('payment_partial'),
  paymentTimeout('payment_timeout'),
  connectError('connect_error'),
  socketConnectionError('socket_connection_error'),
  dataLoadError('data_load_error'),
  qrFetchError('qr_fetch_error'),
  paymentProcessNotAvailable('payment_process_not_available'),
  getAssetsError('get_assets_error'),
  unknownError('unknown_error');

  const ClientCode(this.value);
  final String value;
}

abstract class ClientResponseBase {
  ClientResponseBase({required this.code, required this.message});

  final ClientCode code;
  final String message;
}

class ClientResponse<T extends dynamic> extends ClientResponseBase {
  final T? result;

  ClientResponse({
    required super.code,
    required super.message,
    this.result,
  });

  factory ClientResponse.fromJson(
    Map<String, dynamic> json, {
    String? code,
    String? message,
    T Function(dynamic)? result,
  }) =>
      ClientResponse(
        code: ClientCode.values.firstWhere(
          (code) => code.value == (json['status'] ?? code),
          orElse: () => ClientCode.success,
        ),
        message: json['message'] ?? message ?? I18n.t.successes.success,
        result: result?.call(json['result']) ?? json['result'] ?? json,
      );
}

class ClientError extends ClientResponseBase {
  final dynamic error;

  ClientError({
    required super.code,
    required super.message,
    this.error,
  });

  factory ClientError.fromError(
    dynamic error, {
    ClientCode? code,
    String? message,
  }) {
    if (error is ClientError) return error;

    if (error is Map) {
      return ClientError(
        code: code ??
            ClientCode.values.firstWhere(
              (code) => code.value == (error['statusCode'] ?? error['status']),
              orElse: () => ClientCode.unknownError,
            ),
        message: message ??
            error['message']?.toString() ??
            I18n.t.errors.unknownError,
        error: error['error'] ?? error,
      );
    }
    return ClientError(
      code: code ?? ClientCode.unknownError,
      message: error?.toString() ?? message ?? I18n.t.errors.unknownError,
      error: error,
    );
  }
}
